File: DataScienceFromScratch/python/5.py
#!/usr/local/bin/python2.7

from __future__ import division 
from collections import Counter
import random
import matplotlib
gui_env = ['pgf', 'Qt4Cairo', 'cairo', 'MacOSX', 'TkCairo', 'gdk', 'ps', 'GTKAgg', 'nbAgg', 'GTK', 'Qt5Agg', 'template', 'Qt5Cairo', 'WXCairo', 'GTK3Cairo', 'GTK3Agg', 'WX', 'Qt4Agg', 'TkAgg', 'agg', 'svg', 'GTKCairo', 'WXAgg', 'WebAgg', 'pdf']
matplotlib.use(gui_env[18]) #TkAgg
import matplotlib.pyplot as plt

#############################################################################################
#                                       Statistics 
#############################################################################################

#Base dataset
random.seed(69) #lolz
num_friends = [random.randint(1,100) for x in range(101)]
friend_counts = Counter(num_friends)
xs = range(101)
ys = [friend_counts[x] for x in xs]

#visualization
def prob_view():
    plt.bar(xs,ys)
    plt.axis([0,101,0,25])
    plt.title("Histogram of Friend Counts")
    plt.xlabel("# of friends")
    plt.ylabel("# of people")
    plt.show()
#prob_view()

#Generating some statistics
num_points = len(num_friends)
maximum = max(num_friends)
minimum = min(num_friends)

#Alternatively
sorted_vals = sorted(num_friends)
smallest = sorted_vals[0]
second_smallest = sorted_vals[1]
second_biggest = sorted_vals[-2]

#print "Some Statistics:\nmin = %s\nmax = %s\nsecond largest = %s" % (minimum,maximum,second_biggest)

#################################### Central Tendencies #####################################

#mean (average) -> where is the data centered?
def mean(x):
    return sum(x)/len(x)
#print "mean num of friends: %s" % mean(num_friends)

#median -> middle-most value or average of two middle-most values (# of values is odd/even)
    # good at ignoring outliers 
def median(v):
    """finds middle-most value"""
    n = len(v)
    sorted_v = sorted(v)
    mid = n // 2 #int division

    if n%2 == 1:    #odd
        return sorted_v[mid]
    else:   #even
        lo = mid - 1
        hi = mid
        return (sorted_v[lo]+sorted_v[hi])/2
#print "median num of friends: %s" % median(num_friends)

#represents values less than which a certain percentile of the data lies
def quantile(x, p):
    """returns the pth-percentile value in x"""
    idx = int(p * len(x))
    return sorted(x)[idx]
#print "sample quantiles: 0.1 -> %s , 0.25 -> %s, 0.75 -> %s, 0.9 -> %s" % (quantile(num_friends, 0.1), quantile(num_friends, 0.25), quantile(num_friends, 0.75), quantile(num_friends, 0.90))

#most-common value[s]
def mode(x):
    """returns a list"""
    counts = Counter(x)
    Max = max(counts.values())
    return[x for x, count in counts.iteritems() if count == Max]
print "mode: %s" % mode(num_friends)


#################################### Dispersion #####################################

File: DataScienceFromScratch/julia/5.jl
#!/usr/local/bin/julia
using Pkg
Pkg.activate("jenv")
Pkg.instantiate()

using Plots, Random, StatsBase, DataStructures, LinearAlgebra
gr()

#=

  Chapter 5: Statistics

=#

#dataset
Random.seed!(1234) #reproducability
numFriends = [rand(1:100) for x in 1:100]

#Visualization: Histogram
friendCount = counter(numFriends)
xs = 1:100
ys = [friendCount[x] for x in xs]
#@time bar(xs,ys,xlabel="# of friends",ylabel="# of people",title="Histogram of Friend Counts",show=true,ylims=(0,25))


#Basic Stats
function basic_stats()
  @time numPoints = length(numFriends)
  @time maxVal = max(numFriends...)
  @time minVal = min(numFriends...)
  println("number of points: $numPoints\nmax value: $maxVal\nmin value: $minVal")
end
#basic_stats()


#################################### Central tendencies #####################################
#average
function mean(x)
  return sum(x)/length(x)
end
#println(mean(numFriends))

#middle-most value
function median(x)
  n = length(x)
  sort!(x)
  midpoint = Int(floor(n/2))
  if n%2==1 #odd number values
    return x[midpoint]
  else #even number of values
    lo = midpoint-1
    hi = midpoint
    return (x[lo]+x[hi])/2
  end
end
#println(median(numFriends))

#represents the value less than which a certain percentile of the data lies.
function quantile(x,p)
  idx = Int(p*length(x))
  return sort(x)[idx]
end
#println("quantile @ 10,25,75,90 percent: $(quantile(numFriends,0.1)), $(quantile(numFriends,0.25)), $(quantile(numFriends,0.75)), $(quantile(numFriends,0.9))")

function mode(x)
  counts = countmap(x)
  maxCount = max(collect(values(counts))...)
  println("counts: $counts\nmaxCount: $maxCount")
  return [xi for (xi,count) in counts if count==maxCount]
end
#println(mode(numFriends))


######################################## Dispersion #########################################

#=  measures how spread out data is:
      *values near zero signify not spread out at all
      *large value signify very spread out.
=#

function data_range(x)
  return max(x...)-min(x...)
end
#println("data range: $(data_range(numFriends))")

#Variance

function de_mean(x)
  xbar = mean(x)
  return [xi-xbar for xi in x]
end

sum_of_squares(x) = dot(x,x)

#measures how a single variable deviates from its mean
function variance(x)
  n = length(x)
  deviations = de_mean(x)
  return sum_of_squares(deviations)/(n-1)
end
#println("Variance: $(variance(numFriends))")

function stddev(x)
  return sqrt(variance(x))
end
#println("standard deviation: $(stddev(numFriends)))")


function interquartile_range(x)
  return quantile(x,0.75) - quantile(x,0.25)
end
#println("interquartile range: $(interquartile_range(numFriends))")


######################################## Correlation #########################################

#measures how two variables vary in tandem from their means
function covariance(x,y)
   n = length(x)
   return dot(de_mean(x),de_mean(y))/(n-1)
end

#dailyminutes dataset
dailyMinutes = [rand(1:1000) for x in 1:100]
#println("Covariance: $(covariance(numFriends,dailyMinutes))")


function correlation(x,y)
  stddevx = stddev(x)
  stddevy = stddev(y)
  if stddevx>0 && stddevy>0
    return covariance(x,y)/stddevx/stddevy
  else
    return 0
  end                  
end
println("Correlation: $(correlation(numFriends,dailyMinutes))")


readline() #wait for user input to exit

